File: homework_four_dot_one/src/main/java/com/abra/homework_four_dot_one/MainActivity.java
package com.abra.homework_four_dot_one;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.view.inputmethod.EditorInfo;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.ArrayList;


public class MainActivity extends AppCompatActivity {
    private RecyclerView recyclerView;
    private ContactItemAdapter contactItemAdapter;
    private FloatingActionButton button;
    private SearchView searchView;
    private Toolbar toolbar;
    private static final int ACTIVITY_ADD = 1;
    private static final int ACTIVITY_EDIT = 2;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        toolbar = findViewById(R.id.toolBarSearch);
        setSupportActionBar(toolbar);
        button = findViewById(R.id.floatingActionButton);
        recyclerView = findViewById(R.id.recyclerView);
        searchView = findViewById(R.id.searchView);
        setRecyclerViewSettings(savedInstanceState);
        setSearchViewSettings();
        setFabListener();
    }

    private void setSearchViewSettings() {
        searchView.setImeOptions(EditorInfo.IME_ACTION_DONE);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                contactItemAdapter.getFilter().filter(newText);
                return false;
            }
        });
    }

    private void setRecyclerViewSettings(Bundle savedInstanceState) {
        if (savedInstanceState != null) {
            contactItemAdapter = new ContactItemAdapter(savedInstanceState.getParcelableArrayList("list"));
        } else {
            contactItemAdapter = new ContactItemAdapter(new ArrayList<>());
        }
        int rotation = getWindowManager().getDefaultDisplay().getRotation();
        recyclerView.setAdapter(contactItemAdapter);
        // почему-то работает наоборот, я так и не понял почему
        if (rotation == ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE) {
            recyclerView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));
        } else {
            recyclerView.setLayoutManager(new GridLayoutManager(this, 2));
        }
        contactItemAdapter.setOnItemClickListener(((contactItem, position) -> {
            Intent intent = new Intent(this, EditContactActivity.class);
            intent.putExtra("position", position);
            intent.putExtra("contactItem", contactItem);
            startActivityForResult(intent, ACTIVITY_EDIT);
        }));
    }

    private void setFabListener() {
        button.setOnClickListener(view -> {
            Intent intent = new Intent(this, AddContactActivity.class);
            startActivityForResult(intent, ACTIVITY_ADD);
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (data != null) {
            ContactItem contactItem;
            if (requestCode == ACTIVITY_ADD) {
                contactItem = data.getParcelableExtra("contactItem");
                if (!data.getBooleanExtra("flag", false) && contactItem != null) {
                    contactItemAdapter.add(contactItem);
                }
            } else if (requestCode == ACTIVITY_EDIT) {
                if (!data.getBooleanExtra("isRemoved", false)) {
                    contactItem = data.getParcelableExtra("contactItem");
                    int position = data.getIntExtra("position", 0);
                    if (contactItem != null) contactItemAdapter.edit(position, contactItem);
                } else {
                    contactItemAdapter.remove(data.getIntExtra("position", 0));
                }
            }
        }
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        if (contactItemAdapter.getContactItemList().size() != 0) {
            outState.putParcelableArrayList("list", contactItemAdapter.getContactItemList());
        } else {
            outState.putParcelableArrayList("list", new ArrayList<>());
        }
    }
}

File: homework_7_executor_service/src/main/java/com/abra/homework_7_executor_service/activities/MainActivity.kt
package com.abra.homework_7_executor_service.activities

import android.content.Intent
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.Menu
import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.appcompat.widget.SearchView
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.abra.homework_7_executor_service.R
import com.abra.homework_7_executor_service.adapters.CarInfoAdapter
import com.abra.homework_7_executor_service.data.CarInfo
import com.abra.homework_7_executor_service.repositories.DatabaseRepository
import com.abra.homework_7_executor_service.services.DatabaseService
import com.google.android.material.floatingactionbutton.FloatingActionButton
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date

private const val ADD_ACTIVITY_CODE = 1
private const val EDIT_ACTIVITY_CODE = 2
private const val WORK_LIST_ACTIVITY = 10
private const val RESULT_CODE_BUTTON_BACK = 5

class MainActivity : AppCompatActivity() {
    private lateinit var fab: FloatingActionButton
    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: CarInfoAdapter
    private lateinit var noCarsAddedText: TextView
    private lateinit var searchView: SearchView
    private lateinit var repository: DatabaseRepository
    private lateinit var callbackListener: (List<CarInfo>) -> Unit

    @RequiresApi(Build.VERSION_CODES.P)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        fab = findViewById(R.id.fabAddCar)
        recyclerView = findViewById(R.id.recyclerViewCars)
        noCarsAddedText = findViewById(R.id.tvNoCarsAdded)
        DatabaseRepository.initDatabase(applicationContext)
        repository = DatabaseRepository(DatabaseService.getInstance())
        setRecyclerSettings()
        setFabListener()
        setAdapterListeners()
        writeLogToFile()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.main_menu, menu)
        searchView = menu?.findItem(R.id.search)?.actionView as SearchView
        searchView.apply {
            imeOptions = EditorInfo.IME_ACTION_DONE
            setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(p0: String?) = false

                override fun onQueryTextChange(p0: String?): Boolean {
                    adapter.filter.filter(p0)
                    return false
                }
            })
        }
        return true
    }

    private fun setFabListener() {
        fab.setOnClickListener {
            val intent = Intent(this, AddCarActivity::class.java)
            startActivityForResult(intent, ADD_ACTIVITY_CODE)
        }
    }

    @RequiresApi(Build.VERSION_CODES.P)
    private fun setRecyclerSettings() {
        adapter = CarInfoAdapter()
        recyclerView.adapter = adapter
        recyclerView.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        callbackListener = { list ->
            adapter.updateList(list.sortedBy { it.producer.toLowerCase() })
            setNoCarsTextViewVisibility()
        }
        repository.getAllList(callbackListener, mainExecutor)
    }

    @RequiresApi(Build.VERSION_CODES.P)
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode != RESULT_CODE_BUTTON_BACK) {
            callbackListener = { list ->
                adapter.updateList(list.sortedBy { it.producer.toLowerCase() })
                setNoCarsTextViewVisibility()
            }
            repository.getAllList(callbackListener, mainExecutor)
            if (!searchView.isIconified) {
                searchView.onActionViewCollapsed()
            }
        }
    }

    private fun writeLogToFile() {
        val logList = StringBuilder()
        val file = File(filesDir, "Logs.txt")
        if (!file.exists()) {
            file.createNewFile()
        }
        logList.append("${file.readText()} \n")
        val simpleDateFormat = SimpleDateFormat.getDateTimeInstance()
        val date = simpleDateFormat.format(Date())
        logList.append(date)
        file.writeText(logList.toString())
    }

    private fun setNoCarsTextViewVisibility() {
        if (adapter.itemCount != 0) noCarsAddedText.visibility = View.INVISIBLE
        else noCarsAddedText.visibility = View.VISIBLE
    }

    private fun setAdapterListeners() {
        adapter.onEditIconClickListener = {
            val intent = Intent(this, EditCarInfoActivity::class.java)
            intent.putExtra("carInfo", it)
            startActivityForResult(intent, EDIT_ACTIVITY_CODE)
        }
        adapter.onCarInfoShowWorkListClickListener = {
            val intent = Intent(this, WorkListActivity::class.java)
            intent.putExtra("carInfo", it)
            startActivityForResult(intent, WORK_LIST_ACTIVITY)
        }
    }

    override fun onDestroy() {
        DatabaseService.getInstance().shutdown()
        super.onDestroy()
    }
}